### ISSUE TITLE
BUG: resample inconsistency with closed left/right

### ISSUE THREAD

**Issue by samdoolin** [pandas-dev/pandas#36806]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd

t = pd.date_range("2011-01-01T00:00:00", "2011-02-01T00:00:00", freq="10T")
y = np.ones(len(t))
df = pd.DataFrame(data=y, index=t, columns=['y'])

df

# default for '1M' is closed-left
resampled_df = df.resample('1M', closed='left', label='right').sum()

resampled_df

# bin edges are last day of month @ start of day
# this is why the whole day of 2011-01-31 has been included in the second bin
resampled_df.index[0]

# now with closed-right
# I am expecting 1 data point (precisely on the bin edge @ 2011-01-31 00:00:00) to shift into the first bin
resampled_df = df.resample('1M', closed='right', label='right').sum()

resampled_df

# now the first bin includes the whole day of 2011-01-31
# despite the fact that the bin edge is 2011-01-31 00:00:00
resampled_df.index[0]
```

executed...

```python
>>> import numpy as np
>>> import pandas as pd
>>> 
>>> t = pd.date_range("2011-01-01T00:00:00", "2011-02-01T00:00:00", freq="10T")
>>> y = np.ones(len(t))
>>> df = pd.DataFrame(data=y, index=t, columns=['y'])
>>> 
>>> df
                       y
2011-01-01 00:00:00  1.0
2011-01-01 00:10:00  1.0
2011-01-01 00:20:00  1.0
2011-01-01 00:30:00  1.0
2011-01-01 00:40:00  1.0
...                  ...
2011-01-31 23:20:00  1.0
2011-01-31 23:30:00  1.0
2011-01-31 23:40:00  1.0
2011-01-31 23:50:00  1.0
2011-02-01 00:00:00  1.0

[4465 rows x 1 columns]
>>> 
>>> # default for '1M' is closed-left
>>> resampled_df = df.resample('1M', closed='left', label='right').sum()
>>> 
>>> resampled_df
                 y
2011-01-31  4320.0
2011-02-28   145.0
>>> 
>>> # bin edges are last day of month @ start of day
>>> # this is why the whole day of 2011-01-31 has been included in the second bin
>>> resampled_df.index[0]
Timestamp('2011-01-31 00:00:00', freq='M')
>>> 
>>> # now with closed-right
>>> # I am expecting 1 data point (precisely on the bin edge @ 2011-01-31 00:00:00) to shift into the first bin
>>> resampled_df = df.resample('1M', closed='right', label='right').sum()
>>> 
>>> resampled_df
                 y
2011-01-31  4464.0
2011-02-28     1.0
>>> 
>>> # now the first bin includes the whole day of 2011-01-31
>>> # despite the fact that the bin edge is 2011-01-31 00:00:00
>>> resampled_df.index[0]
Timestamp('2011-01-31 00:00:00', freq='M')
```

#### Problem description

Switching between closed left/right should affect at most 1 data point in a timeseries, but in the example above it affects an entire day of data.

We were also surprised that '1M' (calendar month end) is defined as last day of month _@ start of day_.

#### Expected Output

see comments in example above

#### Output of ``pd.show_versions()``

<details>

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : 2a7d3326dee660824a8433ffd01065f8ac37f7d6
python           : 3.8.1.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.19.76-linuxkit
Version          : #1 SMP Tue May 26 11:42:35 UTC 2020
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.2
numpy            : 1.19.2
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.1.0
Cython           : None
pytest           : 5.3.5
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.3.6
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.4 (dt dec pq3 ext lo64)
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : 0.16.1
xlrd             : None
xlwt             : None
numba            : None
```

</details>

**Comment by samdoolin** [comment]
bug is also present with `.resample('1A', ...)`

**Comment by phofl** [comment]
> We were also surprised that '1M' (calendar month end) is defined as last day of month @ start of day.

I don't think this interpretation is correct. This seems to me as that the calendar month end if interpreted as the last day of the month, period. ``closed=left`` means, that this day is not included into the interval, ``closed=right`` means, that the complete day is included.

If you check the downsampling example in https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#basics (third code box). This is the same behavior with minute.

``ts.resample('5Min', closed='right').mean()``

means, that up until  ``2012-01-01T00:00:59`` is included into ``2011-12-31 23:55:00``

**Comment by joelowj** [comment]
I agree with @phofl 

Using the example above, the 1 month bin range would be 
1. 2010-12-31 to 2011-01-31
2. 2011-01-31 to 2011-02-28

Since label does not impact the results but merely the display, we can stick to `label='right'` for analysis.

If `closed='left'`, then for the 2 cases:
1. 2011-01-31 bin: 2010-12-31 <= date < 2011-01-31
2. 2011-02-28 bin: 2011-01-31 <= date < 2011-02-28

If `closed='right'`, then for the 2 cases:
1. 2011-01-31 bin: 2010-12-31 < date <= 2011-01-31
2. 2011-01-31 bin: 2011-01-31 < date <= 2011-02-28 

When we switch between left and right in your example, there will only be 1 data point in `closed='right'`. This is consistent with the documentation and how I would think of date range; when the range ends at 2011-01-31 then intuitively I would expect every data point for that day up till midnight.

**Comment by samdoolin** [comment]
Thank you phofl and joelowj for your prompt and thoughtful comments.

If I write the bin intervals (as in joelowj's comment), with time included, where a square bracket = closed, and a round bracket = open, then we get...

`.resample('1M', closed='left', label='right')`
1. `[2010-12-31 00:00:00, 2011-01-31 00:00:00)`
2. `[2011-01-31 00:00:00, 2011-02-31 00:00:00)`

`.resample('1M', closed='right', label='right')`
1. `[2010-01-01 00:00:00, 2011-02-01 00:00:00)`
2. `[2011-02-01 00:00:00, 2011-03-01 00:00:00)`

This clearly shows how the bins move to include or not include the last day of the month. If this is indeed to be expected then we must have overlooked something in the documentation.

However, I can argue 2 problems with the `closed='right'` case...

1. the interval is not technically closed-right.
2. the bin right labels on the resampled index are incorrect. e.g. the first bin is labelled as having a right edge at 2011-01-31 00:00:00 but yet includes data up until 2011-02-01 00:00:00

**Comment by joelowj** [comment]
Yes, it looks consistent with how I interpret the parameter `closed` from the documentation. 

Let's use the example you have provided to motivate our discussion. I have slightly modified the date range so that we can analyze 

```
>>> t = pd.date_range("2010-12-30T00:00:00", "2011-02-01T00:00:00", freq="10T")
>>> y = np.ones(len(t))
>>> df = pd.DataFrame(data=y, index=t, columns=['y'])
>>> df.resample('1M', closed='left', label='right').sum()
                 y
2010-12-31   144.0
2011-01-31  4464.0
2011-02-28   145.0
>>> df.resample('1M', closed='right', label='right').sum()
                 y
2010-12-31   288.0
2011-01-31  4464.0
2011-02-28     1.0
```

We can verify the consistency by manually slicing the dataframe based on the date range of the bin intervals which we have agree on. 

If `closed='left'`:
```
>>> len(df[(df.index >= '2010-11-30 00:00:00') & (df.index < '2010-12-31 00:00:00')])
144
>>> len(df[(df.index >= '2010-12-31 00:00:00') & (df.index < '2011-01-31 00:00:00')])
4464
>>> len(df[(df.index >= '2011-01-31 00:00:00') & (df.index < '2011-02-28 00:00:00')])
145
>>>
```

If `closed='right'`:
```
>>> len(df[(df.index >= '2010-12-01 00:00:00') & (df.index < '2011-01-01 00:00:00')])
288
>>> len(df[(df.index >= '2011-01-01 00:00:00') & (df.index < '2011-02-01 00:00:00')])
4464
>>> len(df[(df.index >= '2011-02-01 00:00:00') & (df.index < '2011-03-01 00:00:00')])
1
>>>
```

Again, this is consistent with the results we have gotten from the resample function and what is written in the documentation.  You are right in the observations of how the behavior of `closed` changes with `closed=left` and `closed=right`. But I can't agree that the `closed=right` is not a closed interval. If you picked `closed=right` and resample on a monthly basis, the bin would have included all data points starting the midnight of the first day of the month and ending before the midnight on the first day of the next month (e.g. 2011-01-01 00:00:00 <= date < 2011-02-01 00:00:00). 

Let me quote what is written on the documentation from `closed` parameter,  `The default is ‘left’ for all frequency offsets except for ‘M’, ‘A’, ‘Q’, ‘BM’, ‘BA’, ‘BQ’, and ‘W’ which all have a default of ‘right’.`. This make sense because when we resample on monthly basis, we want the data point for the whole month to be contained in a bin and `closed=right` will give you the whole month of data in the bin. 

If the confusion comes from the label showing `2011-01-31 00:00:00` instead of `2011-01-31 23:59:59` that is understandable. 

```
>>> df.resample('1M', closed='right', label='right').sum().index[1]
Timestamp('2011-01-31 00:00:00', freq='M')
```

This would be a separate issue rather than the function resample being inconsistent. At this point in time, I believe that the behavior of the resample function is consistent with what is written in the documentation.


Happy to hear your thoughts on this.

**Comment by samdoolin** [comment]
thank you again joelowj for a further clear response.

> I can't agree that the closed=right is not a closed interval.

The bin intervals are half-open intervals (closed on one side, open on the other side). The technicality is that a closed-right interval should not contain the left bin edge, but should contain the right bin edge. `(open-left, closed-right]`. However with `.resample('1M', closed='right')` the bin intervals are closed-left / open-right, as we have both confirmed in comments above.

Perhaps we are making a mistake by expecting the `closed` argument to provide such a strong guarantee in all circumstances.

> If the confusion comes from the label showing 2011-01-31 00:00:00 instead of 2011-01-31 23:59:59 that is understandable.

Yes. We were certainly expecting the index of the resampled dataframe to specify the locations of the bin edges precisely.

**Comment by joelowj** [comment]
I can't agree that `close='right'` results in the bin intervals being closed on the left and open on the right. Let me give a more detailed example to motivate the discussion and clarify the misconception.

If we look at an integer set. Given an integer set that are closed on both ends [1, 2, 3, 4, 5]
1. if `closed='left'`, [1, 2, 3, 4, 5) = [1, 2, 3, 4]
2. if `close='right'`, (1, 2, 3, 4, 5] = [2, 3, 4, 5] 

We can generalized this idea to the one we have discussed. Let us introduce a set that contains the range of all dates that are closed on both ends in a given bin (e.g. [2010-12-31, 2010-01-01, ..., 2011-01-31]). 

Note that 
1. if a date is included (e.g. 2011-01-31 then we mean everything from 2011-01-31 00:00:00 to 2011-01-31 23:59:59). 
2. If the date is on the left side of the interval then we refer to it as the beginning at midnight (e.g. [2011-01-31, ...] then 2011-01-31 00:00:00) and if it is on the right side of the interval then we refer to it ending 1 second before a new day (e.g. [ ..., 2011-01-31] then 2011-01-31 23:59:59).

Let's reuse the example we discuss above with standardized notations.

Given `[2010-12-31, 2010-01-01, ..., 2011-01-31]`

1. if `closed='left'`, 
```
[2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31)  =  [2010-12-31, 2010-01-01, ..., 2011-01-30]
[2010-12-31, 2010-01-01, ..., 2011-01-30] = [2010-12-31 00:00:00, ..., 2011-01-30 23:59:59]
```

This is the same as

```
(df.index >= '2010-12-31 00:00:00') & (df.index < '2011-01-31 00:00:00')
```


2. if `closed='right'`, 
```
(2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31] = [2010-01-01, ..., 2011-01-30, 2011-01-31]
[2010-01-01, ..., 2011-01-30, 2011-01-31] = [2010-01-01 00:00:00 , ..., 2011-01-31 23:59:59]
```

This is the same as

```
(df.index >= '2011-01-01 00:00:00') & (df.index < '2011-02-01 00:00:00')
```

To summarize, we can see that given [2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31]
1. if `closed='left'` [2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31)
2. if `closed='right'` (2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31]

When we look at the given set above, we can see that it is not true that the `close='right` is close on the left and open on the right.

**Comment by samdoolin** [comment]
I think that on a discrete index (integer / whole days) the issue is not so clear. However, on a continuous index (float / datetime), we can ask very precisely where the bin interval lies.

In your preceding comment you write

> if `closed='right'` ...  This is the same as `(df.index >= '2011-01-01 00:00:00') & (df.index < '2011-02-01 00:00:00')`

You have defined the lower bin edge as `>=` (including its limit point) (i.e. closed-left) and the upper bin edge as `<` (excluding its limit point) (i.e. open-right).

**Comment by joelowj** [comment]
**When you talked about open or close, you have to reference back to the original set.** In this case, `[2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31]` if `closed='right'` (2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31].

Using this simple example to motivate the discussion, given a set [1, 2, 3, 4, 5] if `closed=right'` then (1, 2, 3, 4, 5] = [2, 3, 4, 5] and `1 < x <= 5` is the same as `2 <= x <= 5`.

Even though you see that the inequality is `>=` for `2011-01-01 00:00:00` and `<` for `2011-02-01 00:00:00` it doesn't determine if the interval is open or close. I could also say change the range to `(df.index > '2011-12-31 23:59:59') & (df.index <= '2011-01-31 23:59:59')` and get the same result. 

How we determine open and closed interval using the original set `[2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31]`:
 - If 2010-12-31 data points are not included in `(df.index >= '2011-01-01 00:00:00') & (df.index < '2011-02-01 00:00:00')` or `(df.index > '2011-12-31 23:59:59') & (df.index <= '2011-01-31 23:59:59')` then this is open on the left. 
 - Similarly if 2011-01-31 data points are in `(df.index >= '2011-01-01 00:00:00') & (df.index < '2011-02-01 00:00:00')` or `(df.index > '2011-12-31 23:59:59') & (df.index <= '2011-01-31 23:59:59')`, it is closed on the right. 

It doesn't matter so much what the inequalities are, what should matter is if the element is included in the set. Hope this clarifies your doubt.

**Comment by samdoolin** [comment]
> When you talked about open or close, you have to reference back to the original set

This was what I was trying to address by pointing out that my input data sits on a "continuous" (not discrete) timeline.

> Even though you see that the inequality is >= for 2011-01-01 00:00:00 and < for 2011-02-01 00:00:00 it doesn't determine if the interval is open or close. I could also say change the range to (df.index > '2011-12-31 23:59:59') & (df.index <= '2011-01-31 23:59:59') and get the same result.

No, you can't. I can always play devil's advocate and invent some new data point (e.g. at 2011-01-31 23:59:59.1) with a fractional amount of seconds (on my continuous timeline) which would then fall into the wrong bin. I maintain that the only way to define the bin interval is `[2011-01-01 00:00:00, 2011-02-01 00:00:00)`.

Please forgive a [wikipedia link](https://en.wikipedia.org/wiki/Interval_(mathematics)#Terminology) that may reveal my ignorance, but prevent us from talking at cross purposes.

**Comment by joelowj** [comment]
@samdoolin, I am clear on what a set interval is. The example were simplified to motivate the discussion and I hope that you are not offended by the simplified example. 

The issue were started because it is claim that there were inconsistency with the resampling function (when we closed on left versus right). I have shown in a few discussion using the example you have provided that the function is consistent with the documentation and I have no reason to believe that there is a bug with the function at this point in time.

Right now the discussion has away from the initially raised issue to the function being unable to handle datetime that goes beyond seconds. Moving forward if you can could up with an example that shows that the current resample function cannot handle datetime that goes beyond `second`, I would be happy to review it further.

**Comment by samdoolin** [comment]
I am not offended, and I thank you for your time in discussing this with me joelowj. I agree that our conversation has become somewhat tangential, and I do not claim that there is a bug in sub-second resampling.

Perhaps it would be productive to simply agree, as we have above, that `.resample('1M', closed='right', label='right')` produces bin right labels on the resampled index that are misleading.

**Comment by jonblower** [comment]
@joelowj Thank you for your responses to this issue, which I must admit is still confusing me, despite your careful explanations. Let me try to explain why the behaviour doesn't match what I would expect, via the simplest example I can think of (based on @samdoolin's above):

Let's create a DataFrame with hourly data from midnight on 1st Jan to midnight on 1st Feb, inclusive. This gives 745 data points (31 days * 24 hours + 1):

```
t = pd.date_range("2011-01-01T00:00:00", "2011-02-01T00:00:00", freq="1H")
y = np.ones(len(t))
df = pd.DataFrame(data=y, index=t, columns=['y'])
```

All but one of those data points fall within the month of January. So if I bin them by month, I would expect to find 744 points in Jan and 1 point in Feb:

```
df.resample("M").count()
```

... and indeed I do, with the default behaviour of `closed='right'`. So far, so good.

It's the behaviour with `closed='left'` that I don't understand. I would expect changing the "closed" behaviour to change the binning of **at most one** data point, i.e. the one that is exactly on the bin boundary. Is that not correct? However:

```
df.resample("M", closed='left').count()
```

This puts 720 data points in the January bin and 25 points in the February bin. So **all the 24 data points from 31st January** are being put into February's bin. This is very surprising to me.

My diagnosis of this is that the following two things are happening:
 - the bin boundaries are set to the **start** of the last day of the month (e.g. `2011-01-31T00:00:00`). This is validated by inspection of the labels.
 - the comparison between data values and bin boundaries is being done **only on the basis of the day of the month**, ignoring the hour of the day.

This is consistent with the observed behaviour for both `closed=right` and `closed=left`.

My expectation would be for the bin boundary to be the last possible instant of the 31st January, which would mean that all January values end up in the January bin.

(In any case, using "MS" as the interval results in more intuitive behaviour - for me at least - as it's easier to define the start of the month than the end. So going forward I would definitely use "MS" but I still think the behaviour for "M" is not expected, unless I've misunderstood something.)

**Comment by joelowj** [comment]
@jonblower @samdoolin just to confirm that you guys understood that M refers to MonthEnd and not Month.

M - MonthEnd
MS - MonthBegin

https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#timeseries-offset-aliases

I just checked that the behavior for M and MS are similar in this case but the range you are looking at for binning are slightly different as the parameter suggest.

**Comment by jonblower** [comment]
@joelowj yes, we have understood the difference between M and MS, thanks.

The question comes down to this: into which bin would you expect `2011-01-31T05:00:00` to be placed in the case of `resample("M", closed='left')`?

I would expect this to end up in the January bin, because it's definitely before the end of the month. It's not on the month boundary. But pandas puts it in the February bin. If you can explain why my expectation is wrong that would be great, although I must say it looks like a bug to me.

**Comment by joelowj** [comment]
@jonblower the explanation is still the same. When you resample by month end, the date range of the bin (assuming closed on both end) is [2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31]. If you `closed='left'`, [2010-12-31, 2010-01-01, ..., 2011-01-30, 2011-01-31), i.e. we don't include any data point on 2011-01-31 in the January bin.

**Comment by jonblower** [comment]
@joelowj OK if that's really the expected behaviour. But "M" is advertised to be month end boundaries. If you're saying that `2011-01-31T05:00:00` is **after the end of January** then I find that very unintuitive, to say the least.

I think the boundaries are not actually month-end but the **start of the last day of the month**, as @samdoolin said. The boundaries are really [2010-12-31T00:00:00, 2011-01-31T00:00:00, ...], when we include the extra precision, right?

**Comment by joelowj** [comment]
if we include the extra precision, it would be `[2010-12-31 00:00:00, 2010-01-01, ..., 2011-01-30, 2011-01-31 23:59:59]`

**Comment by jonblower** [comment]
OK, thank you. I can see that you consider this to be expected behaviour and I understand your explanation about the bin boundaries. I still find it unintuitive from a user point of view that datetimes in the last day of January get sorted into February in this case. I still believe the bin boundaries for "M" ought to be right at the end of the month (e.g. `2010-11-30 23:59:59.999`), not at the start of the last day, but clearly that's not the choice that's been made in pandas. Perhaps it would be worth clarifying the position of the bin boundaries in the pandas docs?

I can also accept that using closed-left with end-month boundaries is an unusual use case, perhaps not to be recommended (which I guess is why the default for "M" is closed-right). And in our case, I think we would be better off using MS boundaries instead anyway, which has exactly the behaviour I would expect. So we have a solution to our issue - we just wanted to flag that we found this particular behaviour unexpected.

Thank you for your time in considering this report.
